extensions/cl_ext_device_side_abort.asciidoc

```diff
@@ -0,0 +1,125 @@
+cl_ext_device_side_abort
+=========================
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:blank: pass:[ +]
+:data-uri:
+:icons: font
+include::../config/attribs.txt[]
+:source-highlighter: coderay
+
+== Name Strings
+
++cl_ext_device_side_abort+
+
+== Contact
+
+Pekka Jääskeläinen, Parmance (pekka /at/ parmance /dot/ com)
+
+== Contributors
+
+Pekka Jääskeläinen, Parmance +
+Henry Linjamäki, Parmance +
+Brice Videau, Argonne National Laboratory
+
+== Notice
+
+Copyright (c) 2021-2022 Parmance and Argonne National Laboratory.
+
+== Status
+
+Draft
+
+== Version
+
+Built On: {docdate} +
+Revision: 3
+
+== Dependencies
+
+This extension is written against the OpenCL Specification Version 1.2, Revision 19.
+
+This extension requires OpenCL 1.2.
+
+== Overview
+
+With the device side abort extension, a work-item (WI) can return from the kernel
+execution at any point and cause abnormal unrecoverable termination of the host
+process.
+
+This extension differs from the cl_arm_controlled_kernel_termination extension
+in that the abort in the device side is expected to behave like a call to the
+POSIX abort() in the host side, terminating also the host process immediately.
+
+== New OpenCL C Types and Functions
+
+[source,c]
+----
+void abort_platform(void);
+----
+
+== Modifications to the OpenCL C Specification
+
+=== Add a new Section 6.12.X - "Device-side invoked platform-wide program abort"
+
+A new function: ::
++
+--
+
+[source,c]
+----
+void abort_platform(void);
+----
+
+Stops the kernel execution and causes an abnormal unrecoverable termination of
+the host process, thus causing abortion of execution of the whole program at
+the platform level.
+
+Calling the device side abort also causes the printf buffer to be flushed on the
+host program's standard output before aborting the host process. This allows to
+implement device side assertion functionality by using the OpenCL C printf
+function and this extension.
+
+The host should abort the process at latest when the kernel command that called
+the function has finished execution, which means that all the WIs will
+either call abort_platform() or return from the kernel command otherwise.
+
+The host process is expected to call the standard abort function of the host
+operating system. In POSIX/C99-compliant systems, this must be implemented by
+a call to abort() which implies that the SIGABRT signal is raised and can be
+caught by defining a signal handler. This means that a minimal implementation
+of a device-side abort with a CPU device can simply call the standard abort()
+directly to implement compliant behavior.
+
+The results of an aborted kernel command are undefined. Device side abort is
+guaranteed to stop the kernel command as soon as possible, even in the presence
+of deadlocks or partially reached barriers in the kernel command.
```

Comment:
Unsure to be honest. I think it's a good idea in principle, but it does change certain implementation details. Like if a thread calling `abort_platform` causes a dead-lock, it might lead to a kernel driver to timeout the GPU job and that alone is undefined behavior, not just "undefined results".

Some implementations might just not be able to terminate the running kernel without undefined behavior. But now this extension actually defines it: it needs to terminate at some point.

I think calling `abort_platform` in itself causes undefined behavior and runtimes can only guarantee anything that happens afterwards on a best effort basis.

Reply:
The basis for this extension's semantics are the one of libc's abort() and the [__trap()](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#miscellaneous-instructions-trap) of CUDA. It would be useful to be able to implement both of those with this extension. 

I'm not sure what CUDA's "generate an interrupt to the host CPU" should imply, but to me the natural semantics here is to abort the whole (host) process and bring down the GPU resources, similar to killing a threaded CPU-only host application with multiple threads running and one of them calling abort().

Reply:
> chipStar implements (or workarounds the lack of) this extension by means of a host-shared variable which a host thread polls and if set to 1 by the device, calls abort(). Do you suggest that even if the process is aborted this way (thus invoking the OS resource cleanup) there is a risk of instability in AMD's case? 

Only if the kernel itself is misbehaving as a result. If the kernel just terminates normally, even after one thread quits, there is no danger. The issue simply just starts once it times out, or causes invalid memory accesses, etc....